src/OpenTelemetry/Metrics/MetricReaderExt.cs

```diff
@@ -0,0 +1,235 @@
+// <copyright file="MetricReaderExt.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Metrics;
+using OpenTelemetry.Internal;
+
+namespace OpenTelemetry.Metrics
+{
+    /// <summary>
+    /// MetricReader which processes individual metrics.
+    /// </summary>
+    public abstract partial class MetricReader
+    {
+        internal const int MaxMetrics = 1000;
+        private readonly Metric[] metrics = new Metric[MaxMetrics];
+        private readonly Metric[] metricsCurrentBatch = new Metric[MaxMetrics];
+        private readonly HashSet<string> metricStreamNames = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+        private readonly object instrumentCreationLock = new object();
+        private int metricIndex = -1;
+
+        internal Metric AddMetricWithNoViews(Instrument instrument)
+        {
+            if (!MeterProviderBuilderSdk.IsValidInstrumentName(instrument.Name))
```

Comment:
Isn't this check is done already by the MeterProviderSdk?

Reply:
It's only done in the `MeterProviderSdk` when there are no views as we just have to look at `Instrument.Name`.

Reply:
yes! But this is the case with no views right?

Reply:
Whoops! I'm not sure what I did wrong here. I remember undoing this commit last night where I have `IsValidInstrumentName` check happening again.

Reply:
Fixed this.